packages/mui-joy/src/Autocomplete/Autocomplete.tsx

```diff
@@ -0,0 +1,899 @@
+  });
+
+  const hasClearIcon = !disableClearable && !disabled && dirty && !readOnly;
+  const hasPopupIcon = (!freeSolo || forcePopupIcon === true) && forcePopupIcon !== false;
+
+  // If you modify this, make sure to keep the `AutocompleteOwnerState` type in sync.
+  const ownerState = {
+    ...props,
+    disablePortal,
+    focused,
+    fullWidth,
+    hasOptions: !!groupedOptions.length,
+    hasClearIcon,
+    hasPopupIcon,
+    inputFocused: focusedTag === -1,
+    popupOpen,
+    size,
+  } as OwnerState;
+
+  const classes = useUtilityClasses(ownerState);
+
+  let startDecorator;
+
+  if (multiple && (value as Array<unknown>).length > 0) {
+    const getCustomizedTagProps: AutocompleteRenderGetTagProps = (params) => {
+      const { onDelete, ...tagProps } = getTagProps(params);
+      return {
+        disabled,
+        size,
+        onClick: onDelete,
+        ...tagProps,
+      };
+    };
+
+    if (renderTags) {
+      startDecorator = renderTags(value as Array<unknown>, getCustomizedTagProps, ownerState);
+    } else {
+      startDecorator = (value as Array<unknown>).map((option, index) => {
+        return (
+          <Chip
+            key={index}
+            size={size}
+            variant="soft"
+            color="neutral"
+            endDecorator={<ChipDelete {...getCustomizedTagProps({ index })} />}
+          >
+            {getOptionLabel(option)}
+          </Chip>
+        );
+      });
+    }
+  }
+
+  const [SlotRoot, rootProps] = useSlot('root', {
+    ref,
+    className: classes.root,
+    elementType: AutocompleteRoot,
+    externalForwardedProps: other,
+    ownerState,
+    getSlotProps: getRootProps,
+  });
+
+  const [SlotClearIndicator, clearIndicatorProps] = useSlot('clearIndicator', {
+    className: classes.clearIndicator,
+    elementType: AutocompleteClearIndicator,
+    getSlotProps: getClearProps,
+    externalForwardedProps: other,
+    ownerState,
+    getSlotOwnerState: (mergedProps) => ({
+      size: mergedProps.size || size,
+      variant: mergedProps.variant || 'plain',
+      color: mergedProps.color || 'neutral',
+    }),
+    additionalProps: {
+      'aria-label': clearText,
+      title: clearText,
+    },
+  });
+
+  const [SlotPopupIndicator, popupIndicatorProps] = useSlot('popupIndicator', {
+    className: classes.popupIndicator,
+    elementType: AutocompletePopupIndicator,
+    getSlotProps: getPopupIndicatorProps,
+    externalForwardedProps: other,
+    ownerState,
+    getSlotOwnerState: (mergedProps) => ({
+      size: mergedProps.size || size,
+      variant: mergedProps.variant || 'plain',
+      color: mergedProps.color || 'neutral',
+    }),
+    additionalProps: {
+      disabled,
+      'aria-label': popupOpen ? closeText : openText,
+      title: popupOpen ? closeText : openText,
+    },
+  });
+
+  // cache the modifiers to prevent Popper from being recreated when React rerenders menu.
+  const cachedModifiers = React.useMemo<PopperUnstyledProps['modifiers']>(
```

Comment:
This could be declared outside the component - there would be no need to call the hook.

Reply:
Done.